feature/manager/src/main/java/com/kinandcarta/create/proxytoggle/manager/viewmodel/ProxyManagerViewModel.kt

```diff
@@ -0,0 +1,212 @@
+        val newTextFiltered = newText
+            .filter(Char::isDigit)
+            .take(ProxyValidator.MAX_PORT.toString().length)
+        updateUiState {
+            it.copy(portState = it.portState.copy(text = newTextFiltered))
+        }
+    }
+
+    fun onForceFocusExecuted() {
+        updateUiState {
+            it.copy(
+                addressState = it.addressState.copy(forceFocus = false),
+                portState = it.portState.copy(forceFocus = false)
+            )
+        }
+    }
+
+    private fun enableProxy() {
+        val address = uiState.value.addressState.text
+        val port = uiState.value.portState.text
+
+        updateErrors(ProxyManagerError.NoError)
+        updateUiState {
+            it.copy(
+                addressState = it.addressState.copy(forceFocus = false),
+                portState = it.portState.copy(forceFocus = false)
+            )
+        }
+        when {
+            !proxyValidator.isValidIP(address) -> {
+                viewModelScope.launch {
+                    delay(ERROR_DELAY)
+                    updateErrors(ProxyManagerError.InvalidAddress)
+                }
+            }
+            !proxyValidator.isValidPort(port) -> {
+                viewModelScope.launch {
+                    delay(ERROR_DELAY)
+                    updateErrors(ProxyManagerError.InvalidPort)
+                }
+            }
+            else -> {
+                deviceSettingsManager.enableProxy(Proxy(address, port))
+            }
+        }
+    }
+
+    private fun updateUiState(updateFunc: (UiState) -> UiState) {
+        _uiState.value = updateFunc(_uiState.value)
+    }
+
+    private fun updateErrors(errorState: ProxyManagerError) {
+        when (errorState) {
+            ProxyManagerError.InvalidAddress -> {
+                updateUiState {
+                    it.copy(
+                        addressState = it.addressState.copy(
+                            error = R.string.error_invalid_address,
+                            forceFocus = true
+                        )
+                    )
+                }
+            }
+            ProxyManagerError.InvalidPort -> {
+                updateUiState {
+                    it.copy(
+                        portState = it.portState.copy(
+                            error = R.string.error_invalid_port,
+                            forceFocus = true
+                        )
+                    )
+                }
+            }
+            ProxyManagerError.NoError -> {
+                updateUiState {
+                    it.copy(
+                        addressState = it.addressState.copy(error = null),
+                        portState = it.portState.copy(error = null)
+                    )
+                }
+            }
+        }
+    }
+
+    private fun getInitialAddressState(): TextFieldState {
+        return TextFieldState(
+            label = R.string.hint_ip_address,
```

Comment:
I think the copy, and things like keyboardOptions are the concern of the View and should not really be part of the ViewModel state. What do you think?

Reply:
IMO this is business logic and I tried to make views as dumb as possible.. idk

Reply:
ok np can move them in there